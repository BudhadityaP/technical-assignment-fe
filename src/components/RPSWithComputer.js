
import React from 'react'
import Button from './Button'
import playRPS, { options, playElements, rpsPropTypes } from './PlayRPS'
import { Container, Row, Col } from 'react-bootstrap';

import userH from '../assets/smiley-happy.png'
import userN from '../assets/smiley-normal.png'
import userL from '../assets/smiley-angry.png'
import userW from '../assets/smiley-winner.png'

import compL from '../assets/Comp-angry.png'
import compW from '../assets/Comp-winner.png'
import compN from '../assets/Comp-normal.png'
import compH from '../assets/Comp-happy.png'

import CommonHeader from './CommonHeader';
import Winner from './Winner'
import RoundResult from './RoundResult'
import { styles } from '../js/styles';

const RPSWithComputer = (props) => {
    const {
        setSelection,
        computerChoice,
        roundResult,
        userChoice,
        restartGame,
        userPoints,
        computerPoints,
        setDropdownValue,
        winner,
        rounds } = props

    return (
        <Container >
            <CommonHeader 
            title={'Play with Computer'}
            setDropdownValue={setDropdownValue}
            rounds={rounds} 
            restartGame={restartGame} />
            
            {winner && <Winner winner={winner}
             userPoints={userPoints} 
             computerPoints={computerPoints}
             page={'pwc'} />}

            <Row style={styles.Container}>
                <Col style={styles.SectionContainer}>
                    {/* Player 1 Section */}
                    <Row style={{ minHeight: 300 }}>
                        <Col>
                            {
                                playElements.map((e, i) => {
                                    return <Row style={{ margin: 5 }} key={e.value}> <Button
                                        isDisabled={winner !== null}
                                        commonProps={styles.common}
                                        color={e.color}
                                        style={styles.Rock}
                                        text={e.name}
                                        onClick={() => setSelection(e.value)}></Button></Row>
                                })
                            }
                        </Col>
                        <Col>
                            {winner ?
                                <Col> {winner === 'p1' ? <img src={userW} alt="User Wins"></img> : <img src={userL} alt="User Loses" />}</Col>
                                : <Col> {roundResult === '1' ? <img src={userH} alt="User Happy"></img> : <img src={userN} alt="User Normal" />}</Col>
                            }
                        </Col>
                    </Row>
                    {userChoice && <Row style={{ alignItems: 'center' }} >
                        <Col><h3>Choice : {userChoice}</h3></Col>
                        <Col >
                            <h3 style={{ ...styles.Points, backgroundColor: '#95BFF4' }}>{userPoints}
                            </h3>
                        </Col>
                    </Row>}

                </Col>
                <Col style={styles.SectionContainer}>
                    {/* Player 2 section */}
                    <Row style={{ minHeight: 300 }}>
                        {
                            winner ?
                                <Col> {winner === 'p2' ? <img src={compW} alt="Comp Wins"></img> : <img src={compL} alt="Comp Loses" />}</Col>
                                : <Col> {roundResult === '2' ? <img src={compH} alt="Comp Happy"></img> : <img src={compN} alt="Comp" />}</Col>
                        }
                    </Row>
                    {computerChoice && <Row style={{ alignItems: 'center' }} >
                        <Col><h3>Choice : {computerChoice}</h3></Col>
                        <Col >
                            <h3 style={{ ...styles.Points, backgroundColor: '#95F49F' }}>{computerPoints}
                            </h3>
                        </Col>
                    </Row>}
                </Col>
            </Row>

            {roundResult && <RoundResult roundResult={roundResult}/>}

        </Container>
    )
}



RPSWithComputer.propTypes = rpsPropTypes
export default playRPS(RPSWithComputer);