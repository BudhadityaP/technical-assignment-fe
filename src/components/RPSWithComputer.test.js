import React from 'react';
import { render, screen, configure, waitFor, fireEvent } from '@testing-library/react';
import RPSWithComputer from './RPSWithComputer';
import { BrowserRouter } from "react-router-dom";

let props = {
    setSelection: jest.fn(),
    computerChoice:'',
    roundResult:'',
    userChoice:'',
    restartGame: jest.fn(),
    userPoints:0,
    computerPoints:0,
    setDropdownValue:jest.fn(),
    rounds:null,
}

describe('----- RPSWithComputer tests -----', () => {
    beforeEach(() => {
        configure({
            throwSuggestions: true,
        })
    })

    it('snapshot test', () => {
        expect(render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)).toMatchSnapshot()
    })

    it('renders correctly - header exists', () => {
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        const h1 = screen.getByRole('heading', { name: /Play with Computer/i })
        const label = screen.getByRole('label', { name: /Please select rounds to play/i })
        expect(h1).toBeDefined();
        expect(label).toBeDefined();
    })

    it('renders correctly - buttons exist', () => {
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        const buttonRestart = screen.getByRole('button', { name: /restart/i })
        const buttonRock = screen.getByRole('button', { name: /rock/i })
        const buttonPaper = screen.getByRole('button', { name: /paper/i })
        const buttonScissor = screen.getByRole('button', { name: /scissor/i })
        expect(buttonRestart).toBeDefined();
        expect(buttonRock).toBeDefined();
        expect(buttonPaper).toBeDefined();
        expect(buttonScissor).toBeDefined();
    })

    it('click restart button', () => {
        render(<BrowserRouter><RPSWithComputer {...props} /></BrowserRouter>)
        const buttonRestart = screen.getByRole('button', { name: /restart/i })
        expect(buttonRestart).toBeDefined();
        fireEvent.click(buttonRestart)
        expect(props.restartGame).toHaveBeenCalled()
    })

    it('click Rock Paper Scissor buttons', () => {
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        const buttonRock = screen.getByRole('button', { name: /rock/i })
        const buttonPaper = screen.getByRole('button', { name: /paper/i })
        const buttonScissor = screen.getByRole('button', { name: /scissor/i })
        fireEvent.click(buttonRock)
        fireEvent.click(buttonPaper)
        fireEvent.click(buttonScissor)
        expect(props.setSelection).toHaveBeenCalledTimes(3)
    })
    
    it('user wins the round', () => {
        props = {...props, roundResult:'1'}
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        expect(screen.getByRole('img', { name: /user happy/i })).toBeDefined()
        expect(screen.getByRole('img', { name: /comp/i })).toBeDefined()
    })

    it('computer wins the round', () => {
        props = {...props, roundResult:'2'}
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        expect(screen.getByRole('img', { name: /comp happy/i })).toBeDefined()
        expect(screen.getByRole('img', { name: /user normal/i })).toBeDefined()
    })

    it('draw game', () => {
        props = {...props, userChoice:'r', computerChoice:'r', roundResult:'Draw'}
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        expect(screen.getByRole('img', { name: /user normal/i })).toBeDefined()
        expect(screen.getByRole('img', { name: /comp/i })).toBeDefined()
    })

    it('user wins the game', () => {
        props = {...props, winner:'p1',roundResult:'1', userChoice:'r', computerChoice:'s'}
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        expect(screen.getByRole('img', { name: /user wins/i })).toBeDefined()
        expect(screen.getByRole('img', { name: /comp loses/i })).toBeDefined()
    })

    it('computer wins the game', () => {
        props = {...props, winner:'p2', roundResult:'2', userChoice:'r', computerChoice:'p'}
        render(<BrowserRouter ><RPSWithComputer {...props} /></BrowserRouter>)
        expect(screen.getByRole('img', { name: /comp wins/i })).toBeDefined()
        expect(screen.getByRole('img', { name: /user loses/i })).toBeDefined()
    })
})

