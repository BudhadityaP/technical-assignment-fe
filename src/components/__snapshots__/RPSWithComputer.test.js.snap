// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`----- RPSWithComputer tests ----- snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="container"
        >
          <div
            class="row"
            style="text-align: center; padding-top: 10px; padding-bottom: 10px;"
          >
            <h1>
              Play with Computer
            </h1>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="row"
                style="align-items: center; text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
              >
                <div
                  class="col"
                >
                  <label
                    role="label"
                    style="font-size: 20px;"
                  >
                    Please select rounds to play 
                  </label>
                </div>
                <div
                  class="col"
                >
                  <div
                    class=" css-b62m3t-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class=" css-1s2u09g-control"
                    >
                      <div
                        class=" css-319lph-ValueContainer"
                      >
                        <div
                          class=" css-14el2xx-placeholder"
                          id="react-select-2-placeholder"
                        >
                          Select...
                        </div>
                        <div
                          class=" css-6j8wv5-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-2-placeholder"
                            aria-expanded="false"
                            aria-haspopup="true"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class=""
                            id="react-select-2-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class=" css-1okebmr-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class=" css-tlfecz-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col"
            >
              <button
                class="btn"
                style="background-color: rgb(181, 196, 203); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px; float: right;"
                type="button"
              >
                Restart
              </button>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
            style="text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
          >
            <div
              class="row"
              style="min-height: 300px;"
            >
              <div
                class="col"
              >
                <div
                  class="row"
                  style="margin: 5px;"
                >
                   
                  <button
                    class="btn"
                    style="background-color: rgb(51, 255, 189); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                    type="button"
                  >
                    Rock
                  </button>
                </div>
                <div
                  class="row"
                  style="margin: 5px;"
                >
                   
                  <button
                    class="btn"
                    style="background-color: rgb(255, 189, 51); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                    type="button"
                  >
                    Paper
                  </button>
                </div>
                <div
                  class="row"
                  style="margin: 5px;"
                >
                   
                  <button
                    class="btn"
                    style="background-color: rgb(51, 167, 255); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                    type="button"
                  >
                    Scissor
                  </button>
                </div>
              </div>
              <div
                class="col"
              >
                <div
                  class="col"
                >
                   
                  <img
                    alt="User Normal"
                    src="[object Object]"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="col"
            style="text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
          >
            <div
              class="row"
              style="min-height: 300px;"
            >
              <div
                class="col"
              >
                 
                <img
                  alt="Comp"
                  src="[object Object]"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="container"
      >
        <div
          class="row"
          style="text-align: center; padding-top: 10px; padding-bottom: 10px;"
        >
          <h1>
            Play with Computer
          </h1>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="row"
              style="align-items: center; text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
            >
              <div
                class="col"
              >
                <label
                  role="label"
                  style="font-size: 20px;"
                >
                  Please select rounds to play 
                </label>
              </div>
              <div
                class="col"
              >
                <div
                  class=" css-b62m3t-container"
                >
                  <span
                    class="css-1f43avz-a11yText-A11yText"
                    id="react-select-2-live-region"
                  />
                  <span
                    aria-atomic="false"
                    aria-live="polite"
                    aria-relevant="additions text"
                    class="css-1f43avz-a11yText-A11yText"
                  />
                  <div
                    class=" css-1s2u09g-control"
                  >
                    <div
                      class=" css-319lph-ValueContainer"
                    >
                      <div
                        class=" css-14el2xx-placeholder"
                        id="react-select-2-placeholder"
                      >
                        Select...
                      </div>
                      <div
                        class=" css-6j8wv5-Input"
                        data-value=""
                      >
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-2-placeholder"
                          aria-expanded="false"
                          aria-haspopup="true"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          class=""
                          id="react-select-2-input"
                          role="combobox"
                          spellcheck="false"
                          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <button
              class="btn"
              style="background-color: rgb(181, 196, 203); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px; float: right;"
              type="button"
            >
              Restart
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
          style="text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
        >
          <div
            class="row"
            style="min-height: 300px;"
          >
            <div
              class="col"
            >
              <div
                class="row"
                style="margin: 5px;"
              >
                 
                <button
                  class="btn"
                  style="background-color: rgb(51, 255, 189); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                  type="button"
                >
                  Rock
                </button>
              </div>
              <div
                class="row"
                style="margin: 5px;"
              >
                 
                <button
                  class="btn"
                  style="background-color: rgb(255, 189, 51); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                  type="button"
                >
                  Paper
                </button>
              </div>
              <div
                class="row"
                style="margin: 5px;"
              >
                 
                <button
                  class="btn"
                  style="background-color: rgb(51, 167, 255); height: 50px; text-align: center; text-decoration: none; font-size: 20px; margin: 20px; width: 200px; border-radius: 10px;"
                  type="button"
                >
                  Scissor
                </button>
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="col"
              >
                 
                <img
                  alt="User Normal"
                  src="[object Object]"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col"
          style="text-align: center; border: 1px solid #808080; border-radius: 10px; margin: 10px; padding: 10px;"
        >
          <div
            class="row"
            style="min-height: 300px;"
          >
            <div
              class="col"
            >
               
              <img
                alt="Comp"
                src="[object Object]"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
